game-core/src/main/java/games/strategy/engine/framework/map/download/DownloadLengthReader.java

```diff
@@ -0,0 +1,103 @@
+package games.strategy.engine.framework.map.download;
+// TODO: move to package games.strategy.engine.framework.map.download.client
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+
+import org.apache.http.Header;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpHead;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.impl.client.CloseableHttpClient;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import games.strategy.engine.framework.system.HttpProxy;
+import lombok.AllArgsConstructor;
+import lombok.extern.java.Log;
+
+/** Can execute an http head request to determine the size of a file download from URI. */
+@Log
+@AllArgsConstructor
+// TODO: testing, break up DownloadUtilsTest to test this component individually from DownloadUtils.
+final class DownloadLengthReader {
+
+  private final Supplier<CloseableHttpClient> httpClientFactory;
+
+  @VisibleForTesting
+  final ConcurrentMap<String, Long> downloadLengthsByUri = new ConcurrentHashMap<>();
+
+  /**
+   * Gets the download length for the resource at the specified URI.
+   *
+   * <p>This method is thread safe.
+   *
+   * @param uri The resource URI; must not be {@code null}.
+   * @return The download length (in bytes) or empty if unknown; never {@code null}.
+   */
+  Optional<Long> getDownloadLength(final String uri) {
+    return getDownloadLengthFromCache(uri, this::getDownloadLengthFromHost);
+  }
+
+  @VisibleForTesting
+  Optional<Long> getDownloadLengthFromCache(
+      final String uri, final DownloadLengthSupplier supplier) {
+    return Optional.ofNullable(
+        downloadLengthsByUri.computeIfAbsent(uri, k -> supplier.get(k).orElse(null)));
+  }
+
+  @VisibleForTesting
+  interface DownloadLengthSupplier {
+    Optional<Long> get(String uri);
+  }
+
+  private Optional<Long> getDownloadLengthFromHost(final String uri) {
+    try (CloseableHttpClient client = httpClientFactory.get()) {
+      return getDownloadLengthFromHost(uri, client);
+    } catch (final IOException e) {
+      log.log(Level.SEVERE, String.format("failed to get download length for '%s'", uri), e);
+      return Optional.empty();
+    }
+  }
+
+  @VisibleForTesting
+  static Optional<Long> getDownloadLengthFromHost(
+      final String uri, final CloseableHttpClient client) throws IOException {
+    try (CloseableHttpResponse response = client.execute(newHttpHeadRequest(uri))) {
```

Comment:
@DanVanAtta Here we're performing a HEAD request to retrieve the download length.